### ISSUE TITLE
babel-node shouldn't pass the arguments-separator (`--`)

### ISSUE THREAD

**Issue by florianb** [babel/babel#5542]
### Input Code
> Example repo: https://github.com/florianb/babelbug

```js
console.dir(process.argv)
```

### Babel Configuration (.babelrc, package.json, cli command)
None

### Expected Behavior
Calling

`babel-node index.js -- --version`

should pass

    [ 'node',
    '/<...>/babelbug/index.js',
    '--version' ]

to the called program.

### Current Behavior
Calling

`babel-node index.js -- --version`

currently passes

    [ 'node',
    '/<...>/babelbug/index.js',
    '--',
    '--version' ]

to the called program.

### Possible Solution
Increase the beginning index of the sliced `userArgs` (in https://github.com/babel/babel/blob/7.0/packages/babel-cli/src/babel-node.js#L17) by one:

```js
userArgs = babelArgs.slice(argSeparator + 1); // including the  --
```

(The comment should, of course, be removed)

### Your Environment

| software         | version
| ---------------- | -------
| Babel            |  
| node             | 
| npm              | 
| Operating System |

**Comment by loganfsmyth** [comment]
I'm open to suggestions here, but I don't know that I agree with your suggestion. `babel-node` at least until now has aimed at essentially being a dropin replacement for the `node` binary, so we pass the arguments straight through.

**Comment by florianb** [comment]
Thanks for your quick response @loganfsmyth - i just finished writing (after having created the issue incidentially).

I guess i have to take further investigations.

**Comment by florianb** [comment]
@loganfsmyth Okay, i guess you're right, but then the problem is even more complicated, i guess. Compare the following calls:

```
~/github/babelbug master
❯ node index.js --version
[ '/Users/.../.nvm/versions/node/v7.4.0/bin/node',
  '/Users/.../github/babelbug/index.js',
  '--version' ]

~/github/babelbug master
❯ babel-node index.js --version
6.18.0
```

It seems like `babel-node` fails at least for the version-option. It looks like babel-node shouldn't process any options behind the script-argument-

**Comment by florianb** [comment]
Probably i should rewrite the issue then? 🤔

**Comment by loganfsmyth** [comment]
Yeah I don't want to imply things are great with the current approach. The cases we'd probably want to handle, which we can explore, are

```
# This argument is passed to Node since it is a V8 option
babel-node --allow_natives_syntax foo.js

# This argument is for node
babel-node --inspect foo.js

# This argument is for babel-node
babel-node --presets es2015 foo.js
```

It seems like the least we can do is require that `babel-node`'s arguments come before the script filename? That way `--version` would pass through to your script properly.

**Comment by loganfsmyth** [comment]
So, find the filename, process all arguments before the filename, pull out the whitelist of `babel-node` arguments, and pass the rest on to Node, and append the rest of the args as args for your script.

**Comment by loganfsmyth** [comment]
Or maybe we take everything up to the first argument we recognize as Node args, then take everything from there up to the filename as babel-node args, and everything after that as script args?

**Comment by florianb** [comment]
Interesting ideas, i guess it would be the easiest to treat everything behind the filename as script-args.

**Comment by florianb** [comment]
But then again, how to distinguish between node-args and babel-args. There aren't that much options overlapping but it may be risky in the future.

**Comment by loganfsmyth** [comment]
That's fair. I don't think there will be a ton of overlap and babel-node doesn't have many options at least. Open to suggestions though.

**Comment by hzoo** [comment]
This sounds like https://github.com/babel/babel/issues/2530? can we close one of them?

**Comment by florianb** [comment]
Great finding @hzoo - i'd close this one if everybody agrees and participate in #2530.

**Comment by florianb** [comment]
Alright @loganfsmyth, i would then propose

## Refine the `babel-node` interface-docs

From currently

    babel-node [options] [ -e script | script.js ] [arguments]

to

    babel-node [options] [script.js | -e "script"] [--] [arguments]

where the `--` should probably be documented as in the [node-docs](https://nodejs.org/api/cli.html#cli):

> Indicate the end of node options. Pass the rest of the arguments to the script. If no script filename or eval/print script is supplied prior to this, then the next argument will be used as a script filename.

## Fix the options parsing

From currently also acting on options after a script-name/eval-option/double-dash, to ignore options after a script-name/eval-option/double-dash.

💡 I would also like to bring to mind that babel-node is hiding node's `--version` and `--help` options.

❓ Should i rewrite the issue-title/-text?

**Comment by jithin1987** [comment]
Another option which is not getting passed by babel-node is --require.

`babel-node --require ./dev/mock.js ./src/script.js` does not work.  It doesn't even run the script.

**Comment by conradoramalho** [comment]
@jimniels how did you fix it?

**Comment by jimniels** [comment]
@conradoramalho I think you meant to tag someone else. I wasn't involved in this discussion at all until you tagged me :)

**Comment by rijnhard** [comment]
Example:

`package.json`
```javascript
{
	"scripts": {
		"console": "node ./dist/console.js",
		"console:dev": "NODE_ENV=development babel-node ./src/console.es"
	}
}
```

Usage: (`-e -s` flags are particularly problematic)
```bash
# working
## raw
NODE_ENV=development ./node_modules/.bin/babel-node ./src/console.es mark -n dev-archive -s 2020-02-15 -e 2020-03-01
## compiled with "--" for npm argument passing
npm run console -- mark -n dev-archive -s "2020-02-15" -e "2020-03-01"

# not working, and no manner of mixing "--" or "--script-args" in run-script or calling works
npm run console:dev -- mark -n dev-archive -s "2020-02-15" -e "2020-03-01"
npm run console:dev mark -n dev-archive -s "2020-02-15" -e "2020-03-01"
# and about 15 other combinations don't work
```

**Comment by jedwards1211** [comment]
@jithin1987 to make matters worse, `--require`/`-r` is undocumented (and I discovered it by accident when trying to run my own CLI app with an `-r` option)

**Comment by nicolo-ribaudo** [comment]
It's a node flag, documented at https://nodejs.org/api/cli.html#cli_r_require_module
